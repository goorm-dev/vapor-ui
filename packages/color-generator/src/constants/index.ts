export * from './colors';
export * from './thresholds';
