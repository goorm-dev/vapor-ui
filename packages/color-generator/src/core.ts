import { formatCss, oklch } from 'culori';

// ============================================================================
// Domain Constants
// ============================================================================

export const DEFAULT_MAIN_BACKGROUND_LIGHTNESS = {
    light: 100,
    dark: 14,
} as const;

export const DEFAULT_CONTRAST_RATIOS = {
    '050': 1.15,
    '100': 1.3,
    '200': 1.7,
    '300': 2.5,
    '400': 3.0,
    '500': 4.5,
    '600': 6.5,
    '700': 8.5,
    '800': 11.5,
    '900': 15.0,
} as const;

// 시맨틱 토큰 카테고리와 변형 정의
export const SEMANTIC_TOKEN_BASE_PATTERNS = {
    Background: 'background',
    Foreground: 'foreground',
    Border: 'border',
    ButtonForeground: 'button-foreground',
} as const;

export const SEMANTIC_TOKEN_VARIANTS = {
    Default: '', // 기본 변형 (background, border, button-foreground)
    Variant100: '100',
    Variant200: '200',
} as const;

// 색상별 시맨틱 토큰 생성 함수 (패턴: {category}-{color}[-{variant}])
export const createSemanticTokenKeys = <T extends string>(colorName: T) => {
    return {
        Background: `background-${colorName}` as const,
        Foreground100: `foreground-${colorName}-100` as const,
        Foreground200: `foreground-${colorName}-200` as const,
        Border: `border-${colorName}` as const,
        ButtonForeground: `button-foreground-${colorName}` as const,
    };
};

// 각 색상별 토큰 키 정의
export const PRIMARY_TOKEN_KEYS = createSemanticTokenKeys('primary');
export const SECONDARY_TOKEN_KEYS = createSemanticTokenKeys('secondary');
export const SUCCESS_TOKEN_KEYS = createSemanticTokenKeys('success');
export const WARNING_TOKEN_KEYS = createSemanticTokenKeys('warning');
export const ERROR_TOKEN_KEYS = createSemanticTokenKeys('error');

// ============================================================================
// Domain Types
// ============================================================================

export type ThemeType = 'light' | 'dark';

export interface ColorToken {
    hex: string;
    oklch: string;
    deltaE?: number;
}

// 일반적인 시맨틱 토큰 맵 타입
export type SemanticDependentTokens<T extends string> = Record<
    ReturnType<typeof createSemanticTokenKeys<T>>[keyof ReturnType<
        typeof createSemanticTokenKeys<T>
    >],
    string
>;

// 각 색상별 토큰 맵 타입
export type PrimaryDependentTokenMap = SemanticDependentTokens<'primary'>;
export type SecondaryDependentTokenMap = SemanticDependentTokens<'secondary'>;
export type SuccessDependentTokenMap = SemanticDependentTokens<'success'>;
export type WarningDependentTokenMap = SemanticDependentTokens<'warning'>;
export type ErrorDependentTokenMap = SemanticDependentTokens<'error'>;

export interface SemanticTokenMap {
    primary: PrimaryDependentTokenMap;
    secondary?: SecondaryDependentTokenMap;
    success?: SuccessDependentTokenMap;
    warning?: WarningDependentTokenMap;
    error?: ErrorDependentTokenMap;
}

export type SemanticColorName = keyof SemanticTokenMap;

export interface ThemeDependentTokensCollection {
    light: SemanticTokenMap;
    dark: SemanticTokenMap;
}

export interface ThemeTokens {
    background: {
        canvas: ColorToken;
    };
    [colorName: string]: {
        [shade: string]: ColorToken;
    };
}

export interface ColorPaletteCollection {
    base: {
        white: ColorToken;
        black: ColorToken;
    };
    light: ThemeTokens;
    dark: ThemeTokens;
}

export interface ColorGeneratorConfig {
    colors?: Record<string, string>;
    contrastRatios?: Record<string, number>;
    backgroundLightness?: {
        light: number;
        dark: number;
    };
}

export interface SemanticColorGeneratorConfig extends ColorGeneratorConfig {
    colors: {
        primary: string;
        secondary?: string;
        success?: string;
        warning?: string;
        error?: string;
        hint?: string;
        contrast?: string;
    };
}

// ============================================================================
// Domain Utilities
// ============================================================================

/**
 * OKLCH 색상 값을 웹 호환성을 위해 포맷팅합니다.
 * - Lightness, Chroma를 3자리 소수점으로 반올림
 * - Hue를 1자리 소수점으로 반올림
 * - 'none' 값을 '0.0'으로 변환
 */
export const formatOklchForWeb = (oklchString: string): string => {
    const match = oklchString.match(/oklch\(([^\s]+)\s+([^\s]+)\s+([^\)]+)\)/);
    if (match) {
        const [, l, c, h] = match;
        const roundedL = parseFloat(l).toFixed(3);
        const roundedC = parseFloat(c).toFixed(3);

        let roundedH: string;
        if (h === 'none' || isNaN(parseFloat(h))) {
            roundedH = '0.0';
        } else {
            roundedH = parseFloat(h).toFixed(1);
        }

        return `oklch(${roundedL} ${roundedC} ${roundedH})`;
    }
    return oklchString;
};

/**
 * Base 컬러 토큰 생성 (흰색, 검은색)
 */
const BASE_COLORS = {
    white: '#ffffff',
    black: '#000000',
} as const;

export const createBaseColorTokens = (formatter: (oklchString: string) => string) => {
    return Object.entries(BASE_COLORS).reduce(
        (tokens, [colorName, hexValue]) => {
            const oklchColor = oklch(hexValue);
            tokens[colorName as keyof typeof BASE_COLORS] = {
                hex: hexValue,
                oklch: formatter(formatCss(oklchColor) ?? ''),
            };
            return tokens;
        },
        {} as Record<keyof typeof BASE_COLORS, ColorToken>,
    );
};
