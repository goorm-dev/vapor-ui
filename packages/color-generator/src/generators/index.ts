export * from './system-color-palette';
export * from './brand-color-palette';
export * from './semantic-mapping';
