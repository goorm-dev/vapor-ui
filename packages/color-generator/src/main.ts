import * as fs from 'fs';
import * as path from 'path';
import { fileURLToPath } from 'url';

import { darkThemeCss, lightThemeCss } from './index';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const dataDir = path.join(__dirname, 'data');
if (!fs.existsSync(dataDir)) {
    fs.mkdirSync(dataDir, { recursive: true });
}

fs.writeFileSync(path.join(dataDir, 'renewal-vapor-color-light-theme.css'), lightThemeCss);
fs.writeFileSync(path.join(dataDir, 'renewal-vapor-color-dark-theme.css'), darkThemeCss);
