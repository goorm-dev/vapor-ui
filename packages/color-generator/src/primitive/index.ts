import { Color } from '@adobe/leonardo-contrast-colors';

import {
    type ColorGeneratorConfig,
    type ColorPaletteCollection,
    DEFAULT_CONTRAST_RATIOS,
    createBaseColorTokens,
    formatOklchForWeb,
} from '../core';
import { createColorDefinition, generateThemeTokens } from '../libs/adobe-leonardo';

// ============================================================================
// Primitive Colors Configuration
// ============================================================================

export const DEFAULT_PRIMITIVE_COLORS = {
    red: '#F5535E',
    pink: '#F26394',
    grape: '#CC5DE8',
    violet: '#8662F3',
    blue: '#448EFE',
    cyan: '#1EBAD2',
    green: '#04A37E',
    lime: '#1EBAD2',
    yellow: '#FFC107',
    orange: '#ED670C',
} as const;

// ============================================================================
// Primitive Color Generation Logic
// ============================================================================

function createVaporColorDefinitions(
    primitiveColors: Record<string, string>,
    contrastRatios: Record<string, number>,
): Color[] {
    return Object.entries(primitiveColors)
        .map(([name, colorHex]) => createColorDefinition({ name, colorHex, contrastRatios }))
        .filter((c): c is Color => c !== null);
}

export function generateColorPalette(config: ColorGeneratorConfig = {}): ColorPaletteCollection {
    const primitiveColors = config.primitiveColors || DEFAULT_PRIMITIVE_COLORS;
    const contrastRatios = config.contrastRatios || DEFAULT_CONTRAST_RATIOS;

    const vaporColorDefinitions = createVaporColorDefinitions(primitiveColors, contrastRatios);

    return {
        base: createBaseColorTokens(formatOklchForWeb),
        light: generateThemeTokens('light', vaporColorDefinitions, config, formatOklchForWeb),
        dark: generateThemeTokens('dark', vaporColorDefinitions, config, formatOklchForWeb),
    };
}
