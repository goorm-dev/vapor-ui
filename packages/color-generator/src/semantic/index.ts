import {
    type ColorGeneratorConfig,
    type ColorPaletteCollection,
    DEFAULT_CONTRAST_RATIOS,
    createBaseColorTokens,
    formatOklchForWeb,
} from '../core';
import { generateThemeTokens } from '../libs/adobe-leonardo';

// ============================================================================
// Semantic Color Configuration
// ============================================================================

export interface SemanticColorConfig extends ColorGeneratorConfig {
    colors: {
        primary: string;
        secondary?: string;
        success?: string;
        warning?: string;
        error?: string;
        hint?: string;
        contrast?: string;
    };
}

// ============================================================================
// Semantic Color Generation Logic
// ============================================================================

/**
 * Generate semantic color palette with primary and optional semantic colors
 */
export function generateSemanticColorPalette(config: SemanticColorConfig): ColorPaletteCollection {
    const contrastRatios = config.contrastRatios || DEFAULT_CONTRAST_RATIOS;

    return {
        base: createBaseColorTokens(formatOklchForWeb),
        light: generateThemeTokens(config.colors, contrastRatios, 'light'),
        dark: generateThemeTokens(config.colors, contrastRatios, 'dark'),
    };
}
