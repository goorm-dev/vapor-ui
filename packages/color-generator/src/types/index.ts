/* -------------------------------------------------------------------------------------------------
 * Core Types
 * -----------------------------------------------------------------------------------------------*/

export type ThemeType = 'light' | 'dark';

export interface OklchColor {
    mode: 'oklch';
    l: number;
    c: number;
    h?: number;
}

export interface ColorToken {
    name?: string;
    hex: string;
    oklch: string;
    deltaE?: number;
    codeSyntax: string;
}

export interface ScaleInfo {
    backgroundScale: string;
    foregroundScale: string;
    alternativeScale: string;
}

/* -------------------------------------------------------------------------------------------------
 * Collection & Configuration Types
 * -----------------------------------------------------------------------------------------------*/

export interface ThemeTokens {
    background: {
        canvas: ColorToken;
    };
    [colorName: string]: {
        [shade: string]: ColorToken;
    };
}

export interface ColorPaletteCollection {
    base: {
        white: ColorToken;
        black: ColorToken;
    };
    light: ThemeTokens;
    dark: ThemeTokens;
}

export interface ColorGeneratorConfig {
    colors?: Record<string, string>;
    contrastRatios?: Record<string, number>;
    backgroundLightness?: {
        light: number;
        dark: number;
    };
}
