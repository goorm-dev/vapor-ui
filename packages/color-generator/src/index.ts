import { generateColorPalette } from './primitive';
import { generateSemanticColorPalette } from './semantic';

// ============================================================================
// Primitive Colors
// ============================================================================

export { generateColorPalette, DEFAULT_PRIMITIVE_COLORS } from './primitive';
export const colorPalette = generateColorPalette();

// ============================================================================
// Semantic Colors
// ============================================================================

export { generateSemanticColorPalette, type SemanticColorConfig } from './semantic';
export const primaryColorPalette = generateSemanticColorPalette({
    colors: {
        primary: '#ffffdc',
    },
});

// ============================================================================
// Core Constants & Types
// ============================================================================
export { DEFAULT_CONTRAST_RATIOS, DEFAULT_MAIN_BACKGROUND_LIGHTNESS } from './core';
export type { ThemeType, ColorToken, ColorPaletteCollection, ColorGeneratorConfig } from './core';
