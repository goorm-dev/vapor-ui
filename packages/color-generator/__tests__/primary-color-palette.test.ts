import { describe, expect, it } from 'vitest';

import { primaryColorPalette } from '../src/index';

describe('Primary Color Generator', () => {
    describe('primaryColorPalette', () => {
        it('should generate consistent primary color palette collection', () => {
            expect(primaryColorPalette).toMatchSnapshot();
        });
    });
});
