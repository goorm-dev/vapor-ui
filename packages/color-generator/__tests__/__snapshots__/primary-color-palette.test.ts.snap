// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Primary Color Generator > primaryColorPalette > should generate consistent primary color palette collection 1`] = `
{
  "base": {
    "black": {
      "hex": "#000000",
      "oklch": "oklch(0.000 0.000 0.0)",
    },
    "white": {
      "hex": "#ffffff",
      "oklch": "oklch(1.000 0.000 0.0)",
    },
  },
  "dark": {
    "background": {
      "canvas": {
        "hex": "#232323",
        "oklch": "oklch(0.256 0.000 0.0)",
      },
    },
    "gray": {
      "050": {
        "deltaE": 0,
        "hex": "#2e2e2e",
        "oklch": "oklch(0.301 0.000 0.0)",
      },
      "100": {
        "deltaE": 0,
        "hex": "#363636",
        "oklch": "oklch(0.333 0.000 0.0)",
      },
      "200": {
        "deltaE": 0,
        "hex": "#474747",
        "oklch": "oklch(0.398 0.000 0.0)",
      },
      "300": {
        "deltaE": 0,
        "hex": "#606060",
        "oklch": "oklch(0.489 0.000 0.0)",
      },
      "400": {
        "deltaE": 0,
        "hex": "#6c6c6c",
        "oklch": "oklch(0.531 0.000 0.0)",
      },
      "500": {
        "deltaE": 0,
        "hex": "#898989",
        "oklch": "oklch(0.630 0.000 0.0)",
      },
      "600": {
        "deltaE": 0,
        "hex": "#a7a7a7",
        "oklch": "oklch(0.728 0.000 0.0)",
      },
      "700": {
        "deltaE": 0,
        "hex": "#bebebe",
        "oklch": "oklch(0.802 0.000 0.0)",
      },
      "800": {
        "deltaE": 0,
        "hex": "#dcdcdc",
        "oklch": "oklch(0.894 0.000 0.0)",
      },
      "900": {
        "deltaE": 0,
        "hex": "#fafafa",
        "oklch": "oklch(0.985 0.000 0.0)",
      },
    },
    "primary": {
      "050": {
        "deltaE": 71.75,
        "hex": "#2f2f17",
        "oklch": "oklch(0.298 0.039 108.6)",
      },
      "100": {
        "deltaE": 66.14,
        "hex": "#383820",
        "oklch": "oklch(0.334 0.039 108.3)",
      },
      "200": {
        "deltaE": 56.76,
        "hex": "#494931",
        "oklch": "oklch(0.398 0.038 108.0)",
      },
      "300": {
        "deltaE": 45.09,
        "hex": "#626248",
        "oklch": "oklch(0.489 0.039 107.8)",
      },
      "400": {
        "deltaE": 40.17,
        "hex": "#6e6e54",
        "oklch": "oklch(0.531 0.039 107.7)",
      },
      "500": {
        "deltaE": 29.67,
        "hex": "#8b8b6f",
        "oklch": "oklch(0.629 0.040 107.5)",
      },
      "600": {
        "deltaE": 20.71,
        "hex": "#a8a88b",
        "oklch": "oklch(0.724 0.040 107.4)",
      },
      "700": {
        "deltaE": 14.2,
        "hex": "#c0c0a1",
        "oklch": "oklch(0.800 0.042 107.3)",
      },
      "800": {
        "deltaE": 6.98,
        "hex": "#dedebd",
        "oklch": "oklch(0.892 0.044 107.3)",
      },
      "900": {
        "deltaE": 0.83,
        "hex": "#fbfbd9",
        "oklch": "oklch(0.979 0.044 107.2)",
      },
    },
  },
  "light": {
    "background": {
      "canvas": {
        "hex": "#ffffff",
        "oklch": "oklch(1.000 0.000 0.0)",
      },
    },
    "gray": {
      "050": {
        "deltaE": 0,
        "hex": "#efefef",
        "oklch": "oklch(0.952 0.000 0.0)",
      },
      "100": {
        "deltaE": 0,
        "hex": "#e1e1e1",
        "oklch": "oklch(0.910 0.000 0.0)",
      },
      "200": {
        "deltaE": 0,
        "hex": "#c6c6c6",
        "oklch": "oklch(0.827 0.000 0.0)",
      },
      "300": {
        "deltaE": 0,
        "hex": "#a3a3a3",
        "oklch": "oklch(0.715 0.000 0.0)",
      },
      "400": {
        "deltaE": 0,
        "hex": "#959595",
        "oklch": "oklch(0.670 0.000 0.0)",
      },
      "500": {
        "deltaE": 0,
        "hex": "#767676",
        "oklch": "oklch(0.566 0.000 0.0)",
      },
      "600": {
        "deltaE": 0,
        "hex": "#5d5d5d",
        "oklch": "oklch(0.478 0.000 0.0)",
      },
      "700": {
        "deltaE": 0,
        "hex": "#4c4c4c",
        "oklch": "oklch(0.417 0.000 0.0)",
      },
      "800": {
        "deltaE": 0,
        "hex": "#393939",
        "oklch": "oklch(0.345 0.000 0.0)",
      },
      "900": {
        "deltaE": 0,
        "hex": "#262626",
        "oklch": "oklch(0.269 0.000 0.0)",
      },
    },
    "primary": {
      "050": {
        "deltaE": 2.86,
        "hex": "#f1f1ce",
        "oklch": "oklch(0.949 0.046 107.2)",
      },
      "100": {
        "deltaE": 5.71,
        "hex": "#e3e4c2",
        "oklch": "oklch(0.909 0.045 108.6)",
      },
      "200": {
        "deltaE": 12.19,
        "hex": "#c8c8a8",
        "oklch": "oklch(0.824 0.043 107.3)",
      },
      "300": {
        "deltaE": 21.35,
        "hex": "#a5a688",
        "oklch": "oklch(0.716 0.042 109.0)",
      },
      "400": {
        "deltaE": 25.87,
        "hex": "#96977a",
        "oklch": "oklch(0.668 0.041 109.1)",
      },
      "500": {
        "deltaE": 36.34,
        "hex": "#78785e",
        "oklch": "oklch(0.565 0.038 107.6)",
      },
      "600": {
        "deltaE": 46.29,
        "hex": "#605f46",
        "oklch": "oklch(0.479 0.038 105.8)",
      },
      "700": {
        "deltaE": 54.24,
        "hex": "#4e4e36",
        "oklch": "oklch(0.417 0.038 107.9)",
      },
      "800": {
        "deltaE": 64.38,
        "hex": "#3b3b23",
        "oklch": "oklch(0.345 0.039 108.3)",
      },
      "900": {
        "deltaE": 76.48,
        "hex": "#282810",
        "oklch": "oklch(0.270 0.039 108.8)",
      },
    },
  },
}
`;
